### PR TITLE
BUG: .unique() on MultiIndex: preserve names

### LINKED ISSUE #20308: BUG: TypeError when calling unique on empty MultiIndex
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd                                   
import numpy as np                                    
x = pd.MultiIndex.from_tuples([(1, 2), (3, 4)], names=['a', 'b'])
y = x[np.array([False, False])] # multiindex y is now empty
y.unique()

-->
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-290d71fd5bc3> in <module>()
      3 x = pd.MultiIndex.from_tuples([(1, 2), (3, 4)]. names=['a', 'b'])
      4 y = x[np.array([False, False])]
----> 5 y.unique()

C:\projects\pandas-dk\pandas\core\indexes\multi.py in unique(self, level)
   1072
   1073         if level is None:
-> 1074             return super(MultiIndex, self).unique()
   1075         else:
   1076             level = self._get_level_number(level)

C:\projects\pandas-dk\pandas\core\indexes\base.py in unique(self, level)
   3971             self._validate_index_level(level)
   3972         result = super(Index, self).unique()
-> 3973         return self._shallow_copy(result)
   3974
   3975     @Appender(base._shared_docs['drop_duplicates'] % _index_doc_kwargs)

C:\projects\pandas-dk\pandas\core\indexes\multi.py in _shallow_copy(self, values, **kwargs)
    559             # discards freq
    560             kwargs.pop('freq', None)
--> 561             return MultiIndex.from_tuples(values, **kwargs)
    562         return self.view()
    563

C:\projects\pandas-dk\pandas\core\indexes\multi.py in from_tuples(cls, tuples, sortorder, names)
   1316             if names is None:
   1317                 msg = 'Cannot infer number of levels from empty list'
-> 1318                 raise TypeError(msg)
   1319             arrays = [[]] * len(names)
   1320         elif isinstance(tuples, (np.ndarray, Index)):

TypeError: Cannot infer number of levels from empty list
```
#### Problem description

Looks similar to #16777. Calling `unique` on an empty single-level index works fine, but doing so on an empty  `MultiIndex` fails. In versions prior to 0.19 `MultiIndex.unique` on an empty `MultiIndex` worked (returning an empty array), but now that the array is being wrapped in an index this is causing a failure.

#### Expected Output
```python
MultiIndex(levels=[[1, 3], [2, 4]],
           labels=[[], []],
           names=['a', 'b'])
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: b6691127523f965003dbf877a358c81af5012989
python: 3.6.4.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 62 Stepping 4, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.0.dev0+457.gb669112
pytest: 3.3.2
pip: 9.0.1
setuptools: 38.4.0
Cython: 0.27.3
numpy: 1.14.0
scipy: 1.0.0
pyarrow: 0.8.0
xarray: 0.10.0
IPython: 6.2.1
sphinx: 1.6.6
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: 0.4.0
matplotlib: 2.1.2
openpyxl: 2.4.10
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.1
pymysql: 0.7.11.None
psycopg2: None
jinja2: 2.10
s3fs: 0.1.2
fastparquet: 0.1.4
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #20568: BUG: unique on an empty MI
```
In [13]: pd.MultiIndex.from_arrays([[], []])
Out[13]: 
MultiIndex(levels=[[], []],
           labels=[[], []])

In [14]: pd.MultiIndex.from_arrays([[], []]).unique()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-14-d5be3b23b8c3> in <module>()
----> 1 pd.MultiIndex.from_arrays([[], []]).unique()

~/pandas/pandas/core/indexes/multi.py in unique(self, level)
   1071 
   1072         if level is None:
-> 1073             return super(MultiIndex, self).unique()
   1074         else:
   1075             level = self._get_level_number(level)

~/pandas/pandas/core/indexes/base.py in unique(self, level)
   4366             self._validate_index_level(level)
   4367         result = super(Index, self).unique()
-> 4368         return self._shallow_copy(result)
   4369 
   4370     def drop_duplicates(self, keep='first'):

~/pandas/pandas/core/indexes/multi.py in _shallow_copy(self, values, **kwargs)
    558             # discards freq
    559             kwargs.pop('freq', None)
--> 560             return MultiIndex.from_tuples(values, **kwargs)
    561         return self.view()
    562 

~/pandas/pandas/core/indexes/multi.py in from_tuples(cls, tuples, sortorder, names)
   1315             if names is None:
   1316                 msg = 'Cannot infer number of levels from empty list'
-> 1317                 raise TypeError(msg)
   1318             arrays = [[]] * len(names)
   1319         elif isinstance(tuples, (np.ndarray, Index)):

TypeError: Cannot infer number of levels from empty list

```

### LINKED ISSUE #20570: .unique() drops names on MultiIndex
#### Code Sample, a copy-pastable example if possible

```python
In [2]: pd.Index([1,2,3], name='a').unique().name
Out[2]: 'a'

In [3]: pd.MultiIndex.from_arrays([[1,2], [1,2]], names=['a', 'b']).unique().names
Out[3]: FrozenList([None, None])
```

#### Problem description

``unique`` preserve ``name`` for flat indexes, it should preserve ``names`` for ``MultiIndex``es..

This causes #20308

#### Expected Output

```python
In [3]: pd.MultiIndex.from_arrays([[1,2], [1,2]], names=['a', 'b']).unique().names
Out[3]: FrozenList(['a', 'b'])
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-6-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.23.0.dev0+719.gf3c4cd670
pytest: 3.5.0
pip: 9.0.1
setuptools: 39.0.1
Cython: 0.25.2
numpy: 1.14.1
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.7.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #20308
- [x] closes #20570
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/test_multi.py [hunk 1]
```diff
@@ -2453,22 +2453,31 @@ def test_get_unique_index(self):
...
-        mi = pd.MultiIndex.from_arrays([list('aaaa'), list('aaaa')])
+        mi = pd.MultiIndex.from_arrays([list('aaaa'), list('aaaa')],
+                                       names=['first', 'second'])
         res = mi.unique()
-        exp = pd.MultiIndex.from_arrays([['a'], ['a']])
+        exp = pd.MultiIndex.from_arrays([['a'], ['a']], names=mi.names)
         tm.assert_index_equal(res, exp)
 
+        # GH #20568 - empty MI
+        mi = pd.MultiIndex.from_arrays([[], []], names=['first', 'second'])
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: can u add a test w/o names as well

**Comment by toobaz** [comment]
@jreback ready to merge if there are no further comments

File: pandas/tests/indexes/test_multi.py [hunk 2]
```diff
@@ -2483,6 +2493,11 @@ def test_unique_level(self, level):
         expected = mi.get_level_values(level)
         tm.assert_index_equal(result, expected)
 
+        # With empty MI
+        mi = pd.MultiIndex.from_arrays([[], []], names=['first', 'second'])
+        result = mi.unique(level=level)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: does this also solve the case where no level is specified?

Comment (toobaz) [reply pandas/tests/indexes/test_multi.py]: No, the [previous test](https://github.com/pandas-dev/pandas/pull/20571/files#diff-be59f06a56750dffb20853248fb3b4e1R2477) does.

**Comment by jreback** [comment]
lgtm. merge away @toobaz

**Pull request merged**